jupyter-org-extensions.el

```diff
@@ -0,0 +1,404 @@
+        (result-start (org-babel-where-is-src-block-result))
+        end)
+    (if result-start
+        (save-excursion
+          (goto-char result-start)
+          (setq end (org-babel-result-end)))
+      (setq end (org-element-property :end src)))
+    (kill-new
+     (buffer-substring
+      (org-element-property :begin src)
+      end))))
+
+;;;###autoload
+(defun jupyter-org-clone-block (&optional below)
+  "Clone the block above the current block.
+
+If BELOW is non-nil, add the cloned block below."
+  (interactive "P")
+  (let* ((src (org-element-context))
+         (code (org-element-property :value src)))
+    (unless (eq (org-element-type src) 'src-block)
+      (error "Not in a source block"))
+    (jupyter-org-insert-src-block below)
+    (delete-char 1)
+    (insert code)
+    ;; move to the end of the last line of the cloned block
+    (forward-line -1)
+    (end-of-line)))
+
+;;;###autoload
+(defun jupyter-org-merge-blocks ()
+  "Merge the current block with the next block."
+  (interactive)
+  (let ((current-src-block (org-element-context))
+        (next-src-block (save-excursion
+                          (org-babel-next-src-block)
+                          (org-element-context))))
+    (unless (eq (org-element-type current-src-block) 'src-block)
+      (error "Not in a source block"))
+    (let ((merged-code (concat (org-element-property :value current-src-block)
+                               (org-element-property :value next-src-block)))
+          (lang (org-element-property :language current-src-block))
+          (switches (or (org-element-property :switches current-src-block) ""))
+          (parameters (or (org-element-property :parameters current-src-block) "")))
+      ;; Remove source blocks
+      (mapc (lambda (src-block)
+              (goto-char (org-element-property :begin src-block))
+              (org-babel-remove-result)
+              (setf (buffer-substring (org-element-property :begin src-block)
+                                      (org-element-property :end src-block))
+                    ""))
+            (list next-src-block current-src-block))
+      ;; Now create the merged block, point is where the current block was
+      (insert
+       (org-element-interpret-data
+        (org-element-put-property
+         (jupyter-org-src-block lang parameters merged-code switches)
+         :post-blank 1)))
+      ;; (insert (format "#+BEGIN_SRC %s %s %s\n%s#+END_SRC\n\n"
+      ;;                 lang switches parameters merged-code))
+      (forward-line -3)
+      (end-of-line))))
+
+;;;###autoload
+(defun jupyter-org-move-src-block (&optional below)
+  "Move source block before of after another.
+
+If BELOW is non-nil, move the block down, otherwise move it up."
+  (interactive)
+  ;;skip all this if there are no previous or next source block
+  (when (condition-case nil
+            (save-excursion
+              (if below
+                  (org-babel-next-src-block)
```

Comment:
What do we want to do about subtree boundaries? `org-babel-{next|previous}-src-block` do not consider subtree boundaries, but does it make sense to move a code block across those boundaries?

For example if we have

```
* A

** B

#+BEGIN_SRC jupyter-python
1 + 1
#+END_SRC

* C

#+BEGIN_SRC jupyter-python
1 + 2
#+END_SRC
```

it is currently possible to move the second source block before the first, but should this be possible? In `org-mode` you can only move subtrees around, e.g. using `org-move-subtree-down`, that are on the same level. I think it would be more consistent with how `org-mode` works to prevent moving source blocks across different subtree levels. Granted source blocks do not have a level like subtrees do.

Reply:
When it comes to using org mode to replicate jupyter notebooks, I'd argue that moving source blocks between boundaries is desired. This is because in jupyter notebooks there is no such concept as a subtree or a boundary.
The current behavior would be more in-line with how notebooks operate.

I also find this functionality more useful. One workflow I use often is to clone a source block (it's written below the current lbock) and then move the new block to a new location, often far from the original block.

What is your take on this?

Reply:
OK, if its similar to the Jupyter notebook. I never really used the notebook features that much, but if people are used to that workflow then it doesn't make sense to not allow it.